Ensure a failing `push!`/`append!` doesn't grow the given `AbstractArray`

Seelengrab opened the pull request:
By doing the conversions first, we can fail early here to prevent growing when the result would not be filled, which should be more correct.

This current version of the PR also contains the same test for `append!`, but that's not (yet) fixed so this is an expected failure in CI and which is the reason why this PR is a draft. I don't really see a way for doing that without allocating lots of intermediaries/a big intermediary first for the conversion, so I wanted to get some feedback before doing that. Also, having this small change run in CI first means I get to see if the `push!` change causes a failure outside of `abstractarrays.jl` and `arrayops.jl`, which are the only two testsuites I've tested locally.

Seelengrab committed:
Ensure a failing `push!` doesn't grow the given `AbstractArray`

Seelengrab committed:
Actually exercise the new code with a minimal array

fredrikekre commented on test/abstractarray.jl:
```diff
@@ -1315,6 +1315,45 @@ Base.pushfirst!(tpa::TestPushArray{T}, a::T) where T = pushfirst!(tpa.data, a)
+    try
+        push!(a, 1)
+    catch
+        @test isempty(a)
```
These need to be outside, or in `finally`, otherwise this test will pass if the `try` block doesn't throw.

Seelengrab replied:
Good point, thanks!

Seelengrab committed:
Incorporate PR feedback about `finally`

Seelengrab committed:
Consolidate tests

Seelengrab committed:
Implement preemptive `convert` checking for `append!`

Seelengrab commented:
So, I've implemented the same kind of logic for `append!`. The testing methodology is a bit.. sparse, so could likely use some more stringent ones, if anyone has an idea for a good way of doing that.

Getting this right is REALLY tricky, because the existing code has all sorts of funny fallback behavior when doing `append!(Any[], ...)` :grimacing: Hard to test for that, since some parts of the test suite rely on that working correctly in the first place...

Since this is quite an important piece of code, I'm worried about any potential performance impact this may have, especially since the "unhappy" case now allocates an intermediary. I think there was some effects optimizations here earlier, which the core routine still keeps, but I don't know how bad the other functions are in this regard. Some review by @aviatesk in regards to that would be appreciated. From a cursory look at julia build times while developing this, it shouldn't have a negative impact.

giordano commented:
I thought this had been recently reported in an issue, but I can't find. If anyone can retrieve it, please link it to this PR

Seelengrab commented:
Yeah, if that issue exists, would be good to link it - I only took a look due to @mfiano mentioning it on zulip.

Seelengrab commented:
Well, looks like my gut feeling was correct - there are some [effects failures](https://buildkite.com/julialang/julia-master/builds/24145#0188392b-dbba-41c1-b6c9-6a0a1d0a8d1b/759-822):

https://github.com/JuliaLang/julia/blob/d2f5bbd7cfbac902db952b465b83d242efcf6f08/test/compiler/effects.jl#L835-L845

I'll have to look into what exactly was the bad change here.

Seelengrab committed:
Fix effects tests by adding a specialization for tuples

Seelengrab commented on base/array.jl:
```diff
@@ -1169,24 +1171,71 @@ and [`prepend!`](@ref) and [`pushfirst!`](@ref) for the opposite order.
 
-function append!(a::Vector, items::AbstractVector)
+append!(a::AbstractVector, iter) = _append!(a, IteratorSize(iter), iter)
+append!(a::AbstractVector{T}, it1::T, it2::T, iter::Vararg{T}) where T = foldl(_append_converted!, (it1, it2, iter...); init=a)
```
This is intended for things like `append!(Int[], 1, 2)` (which was legal in the original code), but is technically bad/wrong, because `it1` and `it2` have the eltype of the container, and are not actually containers themselves. I think this could/should be removed as a bug, but that likely needs a PkgEval.

aviatesk replied:
What isn't this like `append!(a::AbstractVector{T}, it::T, iter::Vararg{T}) where T`?

Seelengrab replied:
Good question - that's likely a leftover from before I added the tuple method. I seem to remember there was also some weird test failure. I'll change it and see if the tests run through!

Seelengrab commented:
The fix for the effects test (at least locally) was simple - just make sure tuples don't lead to allocations. Still, there may be other effects bugs hiding that the tests havent caught so far :shrug:

Seelengrab commented:
32 bit doesn't like some `ccall`:

> ```
> llvmcall                                         (17) \|        started at 2023-05-20T15:44:43.736
> --
>   | From worker 13:	fatal: error thrown and no exception handler available.
>   | From worker 13:	ErrorException("fatal error allocating signal stack: mmap: Cannot allocate memory")
>   | From worker 13:	ijl_errorf at /cache/build/default-amdci5-2/julialang/julia-master/src/rtutils.c:77
>   | From worker 13:	alloc_sigstack at /cache/build/default-amdci5-2/julialang/julia-master/src/signals-unix.c:625 [inlined]
>   | From worker 13:	jl_install_thread_signal_handler at /cache/build/default-amdci5-2/julialang/julia-master/src/signals-unix.c:632
>   | From worker 13:	jl_init_root_task at /cache/build/default-amdci5-2/julialang/julia-master/src/task.c:1723
>   | From worker 13:	ijl_adopt_thread at /cache/build/default-amdci5-2/julialang/julia-master/src/threading.c:418
>   | From worker 13:	unknown function (ip: 0xb7d4340b)
>   | From worker 13:	jl_work_wrapper at /cache/build/default-amdci5-2/julialang/julia-master/src/jl_uv.c:1045
>   | From worker 13:	uv__queue_work at /workspace/srcdir/libuv/src/threadpool.c:305
>   | From worker 13:	worker at /workspace/srcdir/libuv/src/threadpool.c:122
>   | From worker 13:	start_thread at /lib/i386-linux-gnu/libpthread.so.0 (unknown line)
>   | From worker 13:	clone at /lib/i386-linux-gnu/libc.so.6 (unknown line)
> ```

and FreeBSD doesn't like floating point:

> ```
> Error During Test at /usr/home/julia/.buildkite-agent/builds/freebsd12-amdci6-0/julialang/julia-master/julia-379c2dfc57/share/julia/test/rounding.jl:246
> --
>   | Test threw exception
>   | Expression: round(Ti, nextfloat(Tf(-0.5)), RoundNearestTiesUp) == 0
>   | InexactError: trunc(UInt64, -0.0)
>   | Stacktrace:
> ```

I don't think these are caused by this PR :eyes:

fredrikekre commented:
> I thought this had been recently reported in an issue, but I can't find. If anyone can retrieve it, please link it to this PR

#15868, #47391

Seelengrab commented:
Thank you! I didn't noticec `prepend!` having the same issue, and it looks like the PR by @MasonProtter just kind of stalled? I can add the same kind of fix for `prepend!` here too (preserving effects of course), but I don't particularly mind which ultimately gets merged :shrug: 

It's just a shame that this has been open (with PRs closing it!) for *7* years now..

giordano commented:
> Talk about recent :(

I think I saw @MasonProtter PR, I actually searched issues opened by him because I thought he reported the issue, didn't think about searching PRs instead.

aviatesk committed:
Merge branch 'master' into push_convert_fix

aviatesk commented:
@nanosoldier `runbenchmarks("array" || "broadcast" || "collection" || "simd" || "sparse" || "tuple", vs=":master")`

aviatesk commented:
Let's see if there are any performance regression detected in the base benchmark.

Seelengrab commented:
Yeah, the `union` benchmarks being hit here is unsurprising - there is an additional allocation & copy here after all, which that code always hits. I'm a bit surprised by the heavy hit to the `simd` benchmarks though - where can I look at the code for them, so I can try to fix that locally?

It'd be possible to avoid the copy if we knew whether `convert` on any of the possible element types may throw an error, which would lead to the intermediate only being necessary when getting a container with eltype `Any` (otherwise we can throw an error early?).

Hard to say what's best here - whether this should error or not and/or should corrupt the mutable container or not is a bit controversial, looking at the previous discussion.. I personally think that detectable errors should be prevented from corrupting incoming arguments, simply to make even erronous use a tiny bit safer :shrug:

aviatesk commented:
The observed difference appears to be within the range of noise. To verify, I will rerun the benchmarks.
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

> where can I look at the code for them, so I can try to fix that locally?

You can find its implementation at the following location:
https://github.com/JuliaCI/BaseBenchmarks.jl/tree/nanosoldier/src

MasonProtter commented:
Apologies, I got stuck on that PR long enough that I forgot to come back to it.

@Seelengrab any thoughts on the relative advantages and disadvantages of the approach you’ve taken here versus the one I took?

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/49905#issuecomment-1557039727) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/5a4314a_vs_672f609/report.md).

Seelengrab commented:
@MasonProtter I think the main differences are that the approach in this PR, while potentially costing an allocation in the "worst" case of not knowing whether `convert` succeeds, is that there's no `try/catch` to "fix" the erronous resize, which may mean a cold cache, since the array may have been reallocated unnecessarily. There's probably also weird interactions with other threads erronously accessing the `array` before it's been done resizing, asking for its `length` while it's being (erronously) resized, only to be met with inconsistent state/writes when the size changes again.

So I think only changing the input once we *know* that it will ultimately be in a consistent state in terms of `convert` succeeding is a good thing.

aviatesk commented:
The benchmark looks okay. Any update on this [comment](https://github.com/JuliaLang/julia/pull/49905#discussion_r1201127941)? And are there any other considerations that we should address before merging?

Seelengrab committed:
Simplfy slurping `append!` signature

Seelengrab committed:
Ensure `prepend!` doesn't change the incoming array if we can know that it could fail

Seelengrab committed:
Test non-growing failure case for `pushfirst!` and `prepend!`

Seelengrab committed:
Test `append!` & similar with OffsetArray as well

Seelengrab commented:
I've addressed the comment and implemented the same fix for `prepend!`, as well as adding the `OffsetArray`-style test from @MasonProtter's PR -  CI should be happy about it, as a far as I can tell from local testing. We can run nanosoldier again for the `prepend!` changes if you want, but other than that, I'm fine with merging.

Seelengrab commented:
> append!(a::AbstractVector{T}, it::T, iter::T...) where T @ Core.Compiler array.jl:1175 === append!B::Core.Compiler.BitArray{1}, items) @ Core.Compiler bitarray.jl:786

Well, that also explains it - the additional `it2` was for ambiguity prevention :facepalm: Fix is coming up.

Seelengrab committed:
Fix ambiguity tests

Seelengrab commented:
The failure is a timeout:

```
Error in testset cmdlineargs:
--
  | Test Failed at /cache/build/default-amdci4-4/julialang/julia-master/julia-71c7e0290f/share/julia/test/cmdlineargs.jl:339
  | Expression: readchomp(`$exename -q -p 2 -e "println(nworkers())"`) == "2"
  | Evaluated: "      From worker 3:\tMaster process (id 1) could not connect within 60.0 seconds.\n      From worker 3:\texiting.\n2" == "2"
```

So seems unrelated.
